File: 53_Maximum Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # The max subarray sum can be viewed as the height difference between bottom and peak.
        # Therefore, at each node, we can either +presum or restart as presum
        # However, if presum <= 0, the sum is always greater to restart;
        #          if presum > 0, the sum is always greater to continue to add.
        # Edge Case: all negative components
        
        # Prefix Sum + Greedy
        res, curSum = -inf, 0
        # Case: all negative || non-negative
        for n in nums:
            # case: single digit
            res = max(res, n)
            # case: subarray
            curSum += n
            if curSum < 0: curSum = 0
            else: res = max(res, curSum)

        return res

# # Greedy + DP
# class Solution:
#     def maxSubArray(self, nums: List[int]) -> int:
#         # For nums[i], either continue to add or restart from itself
#         # except the edge case of all negative, once negative -> discard
#         res, curSum = -inf, 0
#         for i in nums:
#             res = max(res, curSum + i, i)
#             if curSum + i >= 0: curSum += i
#             else: curSum = 0
#         return res

File: 53_Maximum Subarray.java
// 时间复杂度：O(n)，其中 n 为 nums 的长度。
// 空间复杂度：O(1)。仅用到若干额外变量。

class Solution {
    public int maxSubArray(int[] nums) {
        int res = nums[0], presum = nums[0];

        for (int i = 1; i < nums.length; i++) {
            presum = Math.max(nums[i], nums[i] + presum);
            res = Math.max(res, presum);
        }

        return res;
    }
}